File: learning-akka/examples/src/main/java/book/java/ch2/ReverseStringActor.java
package book.java.ch2;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.japi.pf.ReceiveBuilder;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

public class ReverseStringActor extends AbstractActor {

	@Override
	public PartialFunction<Object, BoxedUnit> receive() {
		return ReceiveBuilder.
				match(String.class, s -> sender().tell(new Status.Success(reverse(s)), self())).
				matchAny(o -> sender().tell(new Status.Failure(new ClassNotFoundException()), self()))
				.build();
	}

	private static String reverse(String str) {
		StringBuilder sb = new StringBuilder(str.length());
		for (int i = str.length() - 1; i >= 0; i--)
			sb.append(str.charAt(i));
		return sb.toString();
	}

}

File: learning-akka/examples/src/main/kotlin/book/kotlin/ch2/ReverseStringActor.kt
package book.kotlin.ch2

import akka.actor.Status
import book.kotlin.util.KotlinActor

class ReverseStringActor : KotlinActor() {

	override fun onReceive(message: Any) = when (message) {
		is String -> sender.tell(message.reversed())
		else -> sender.tell(Status.Failure(ClassNotFoundException()))
	}

}

